mobile/salto/src-tauri/src/commands/get_prices.rs

```diff
@@ -0,0 +1,131 @@
+use std::{env, time::Duration};
+use tauri::{AppHandle, Emitter, async_runtime};
+
+use crate::AppState;
+use crate::PriceConfig;
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Token {
+    currency: String,
+    value: f64,
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Price {
+    symbol: String,
+    price: Vec<Token>,
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct PriceResponce {
+    prices: Vec<Price>,
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CurrenciesResponce {
+    currencies: Vec<String>,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error(transparent)]
+    Reqwest(#[from] reqwest::Error),
+    #[error(transparent)]
+    Json(#[from] serde_json::Error),
+    #[error("mutex lock error: {0}")]
+    Lock(String),
+}
+
+impl serde::Serialize for Error {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(self.to_string().as_ref())
+    }
+}
+
+#[tauri::command]
+pub async fn update_get_prices_config(
```

Comment:
The front should be able to set the current currency, we want the user to be able to do that from the config page.

But the assets should be deduced from the tokens present in the wallet at the time of calling the server.

Also, when a wad is received, if it contains an unknown asset. We should call the server asap and to get this new value.

Reply:
I think a good way to design this is https://en.wikipedia.org/wiki/Agent-oriented_programming

Basically:
- spawn a tokio task that will call the server every X seconds
- when receiving the response, it will emit an event that are listened to by the front-end in order to update its own state
- automatically handle retries in case of errors (don't just kill it)
- This task also listens to a specific event that orders him to fire a new request for a specific asset that was just received. When it receive it, it will call the server, and then emit an event for the front end to update it's own state